src/Neo/SmartContract/Native/TokenManagement.NonFungible.cs

```diff
@@ -0,0 +1,289 @@
+    /// <param name="engine">The current <see cref="ApplicationEngine"/> instance.</param>
+    /// <param name="uniqueId">The unique id of the NFT to burn.</param>
+    /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+    /// <exception cref="InvalidOperationException">If the unique id does not exist or owner has insufficient balance or caller is not owner contract.</exception>
+    [ContractMethod(CpuFee = 1 << 17, RequiredCallFlags = CallFlags.All)]
+    internal async Task BurnNFT(ApplicationEngine engine, UInt160 uniqueId)
+    {
+        StorageKey key = CreateStorageKey(Prefix_NFTState, uniqueId);
+        NFTState nft = engine.SnapshotCache.TryGet(key)?.GetInteroperable<NFTState>()
+            ?? throw new InvalidOperationException("The unique id does not exist.");
+        AddTotalSupply(engine, TokenType.NonFungible, nft.AssetId, BigInteger.MinusOne, assertOwner: true);
+        if (!AddBalance(engine.SnapshotCache, nft.AssetId, nft.Owner, BigInteger.MinusOne))
+            throw new InvalidOperationException("Insufficient balance to burn.");
+        engine.SnapshotCache.Delete(key);
+        key = CreateStorageKey(Prefix_NFTAssetIdUniqueIdIndex, nft.AssetId, uniqueId);
+        engine.SnapshotCache.Delete(key);
+        key = CreateStorageKey(Prefix_NFTOwnerUniqueIdIndex, nft.Owner, uniqueId);
+        engine.SnapshotCache.Delete(key);
+        await PostNFTTransferAsync(engine, uniqueId, nft.Owner, null, StackItem.Null, callOnPayment: false);
+    }
+
+    /// <summary>
+    /// Transfers an NFT between owners.
+    /// </summary>
+    /// <param name="engine">The current <see cref="ApplicationEngine"/> instance.</param>
+    /// <param name="uniqueId">The unique id of the NFT.</param>
+    /// <param name="from">The current owner account <see cref="UInt160"/>.</param>
+    /// <param name="to">The recipient account <see cref="UInt160"/>.</param>
+    /// <param name="data">Arbitrary data passed to <c>onNFTPayment</c> or <c>onNFTTransfer</c> callbacks.</param>
+    /// <returns><c>true</c> if the transfer succeeded; otherwise <c>false</c>.</returns>
+    /// <exception cref="InvalidOperationException">If the unique id does not exist.</exception>
+    [ContractMethod(CpuFee = 1 << 17, StorageFee = 1 << 7, RequiredCallFlags = CallFlags.All)]
+    internal async Task<bool> TransferNFT(ApplicationEngine engine, UInt160 uniqueId, UInt160 from, UInt160 to, StackItem data)
+    {
+        StorageKey key_nft = CreateStorageKey(Prefix_NFTState, uniqueId);
+        NFTState nft = engine.SnapshotCache.TryGet(key_nft)?.GetInteroperable<NFTState>()
+            ?? throw new InvalidOperationException("The unique id does not exist.");
+        if (nft.Owner != from) return false;
+        if (!engine.CheckWitnessInternal(from)) return false;
+        StorageKey key = CreateStorageKey(Prefix_TokenState, nft.AssetId);
+        TokenState token = engine.SnapshotCache.TryGet(key)!.GetInteroperable<TokenState>();
```

Comment:
```suggestion
        TokenState token = engine.SnapshotCache.TryGet(key)!.GetInteroperable<TokenState>();
        if (token.Type != TokenType.NonFungible)
            throw new InvalidOperationException("The asset id and the token type do not match");
```
Should not add a check like the fungible case?